dashboard/app/models/user.rb

```diff
@@ -502,8 +511,8 @@ def user_progress_by_stage(stage)
   end
 
   def user_level_for(script_level, level)
-    user_levels.find_by(script_id: script_level.script_id,
-                        level_id: level.id)
+    user_levels_for_script = cached_user_levels_by_script(script_level.script_id)
+    user_levels_for_script[level.id]
```

Comment:
In my world view, `cached_user_levels_by_script` should be `cache_find_user_level(level_id, script_id = nil, fetch_entire_script:)` or some such, where we preemptively populate the cache with all user_levels in `script_id` if `fetch_entire_script`.